shared-bindings/_protomatter/Protomatter.c

```diff
@@ -0,0 +1,293 @@
+    protomatter_protomatter_obj_t *self = (protomatter_protomatter_obj_t*)self_in;
+    check_for_deinit(self);
+    return mp_obj_new_bool(self->paused);
+}
+MP_DEFINE_CONST_FUN_OBJ_1(protomatter_protomatter_get_paused_obj, protomatter_protomatter_get_paused);
+
+STATIC mp_obj_t protomatter_protomatter_set_paused(mp_obj_t self_in, mp_obj_t value_in)  {
+    protomatter_protomatter_obj_t *self = (protomatter_protomatter_obj_t*)self_in;
+    check_for_deinit(self);
+    bool paused = mp_obj_is_true(value_in);
+    if (paused && !self->paused) {
+        _PM_stop(&self->core);
+    } else if (!paused && self->paused) {
+        _PM_resume(&self->core);
+    }
+    self->paused = paused;
+
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_2(protomatter_protomatter_set_paused_obj, protomatter_protomatter_set_paused);
+
+const mp_obj_property_t protomatter_protomatter_paused_obj = {
+    .base.type = &mp_type_property,
+    .proxy = {(mp_obj_t)&protomatter_protomatter_get_paused_obj,
+              (mp_obj_t)&protomatter_protomatter_set_paused_obj,
+              (mp_obj_t)&mp_const_none_obj},
+};
+
+//|   .. method:: write(buf)
+//|
+//|     Transmits the color data in the buffer to the pixels so that they are shown.
+//|
+//|     The data in the buffer must be in "RGB565" format.  This means
+//|     that it is organized as a series of 16-bit numbers where the highest 5
+//|     bits are interpreted as red, the next 6 as green, and the final 5 as
+//|     blue.  The object can be any buffer, but `array.array` and `ulab.array`
+//|     objects are most often useful.
+//|
+STATIC mp_obj_t protomatter_protomatter_swapbuffers(mp_obj_t self_in) {
+    protomatter_protomatter_obj_t *self = (protomatter_protomatter_obj_t*)self_in;
+    check_for_deinit(self);
+
+    _PM_convert_565(&self->core, self->bufinfo.buf, self->width);
+    _PM_swapbuffer_maybe(&self->core);
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_1(protomatter_protomatter_swapbuffers_obj, protomatter_protomatter_swapbuffers);
+
+STATIC const mp_rom_map_elem_t protomatter_protomatter_locals_dict_table[] = {
+    { MP_ROM_QSTR(MP_QSTR_deinit), MP_ROM_PTR(&protomatter_protomatter_deinit_obj) },
+    { MP_ROM_QSTR(MP_QSTR_paused), MP_ROM_PTR(&protomatter_protomatter_paused_obj) },
+    { MP_ROM_QSTR(MP_QSTR_swapbuffers), MP_ROM_PTR(&protomatter_protomatter_swapbuffers_obj) },
+};
+STATIC MP_DEFINE_CONST_DICT(protomatter_protomatter_locals_dict, protomatter_protomatter_locals_dict_table);
+
+STATIC void protomatter_protomatter_get_bufinfo(mp_obj_t self_in, mp_buffer_info_t *bufinfo) {
+    protomatter_protomatter_obj_t *self = (protomatter_protomatter_obj_t*)self_in;
+    check_for_deinit(self);
+    
+    *bufinfo = self->bufinfo;
+}
+
+STATIC void protomatter_protomatter_swapbuffers_void(mp_obj_t self_in) {
+    protomatter_protomatter_swapbuffers(self_in);
+}
+
+STATIC void protomatter_protomatter_deinit_void(mp_obj_t self_in) {
```

Comment:
Why have these `_void` variants?

Reply:
The Python API versions must return mp_obj_t, but the protocol versions don't deal in objects so they return void.

Reply:
This sounds similar to what common_hal is used for, C data structures rather than Python ones. Could you use the common_hal variants in the protocol?

These wrapper functions don't make much sense to me because the user of the protocol could just throw away the return too.